documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/WebSocketDocs.java

```diff
@@ -0,0 +1,491 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.docs.programming;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.time.Duration;
+
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.websocket.api.RemoteEndpoint;
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.StatusCode;
+import org.eclipse.jetty.websocket.api.WebSocketListener;
+import org.eclipse.jetty.websocket.api.WebSocketPartialListener;
+import org.eclipse.jetty.websocket.api.WebSocketPingPongListener;
+import org.eclipse.jetty.websocket.api.WriteCallback;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketConnect;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketError;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
+import org.eclipse.jetty.websocket.api.annotations.WebSocket;
+
+@SuppressWarnings("unused")
+public class WebSocketDocs
+{
+    @SuppressWarnings("InnerClassMayBeStatic")
+    // tag::listenerEndpoint[]
+    public class ListenerEndPoint implements WebSocketListener // <1>
+    {
+        private Session session;
+
+        @Override
+        public void onWebSocketConnect(Session session)
+        {
+            // The WebSocket connection is established.
+
+            // Store the session to be able to send data to the remote peer.
+            this.session = session;
+
+            // You may configure the session.
+            session.setMaxTextMessageSize(16 * 1024);
+
+            // You may immediately send a message to the remote peer.
+            session.getRemote().sendString("connected", WriteCallback.NOOP);
+        }
+
+        @Override
+        public void onWebSocketClose(int statusCode, String reason)
+        {
+            // The WebSocket connection is closed.
```

Comment:
That is so messy. How can an application know if "we have not sent a close frame". 
What does "fully closed" mean? Also below you mention that if the connection has been closed _without_ receiving a CLOSE frame, then this is invoked?!?
How can an application distinguish all these cases?
I'd rather document "you can only do this one thing" rather than "this method is invoked in these many cases, but you cannot tell the difference among them, and depending on each case you may or may not do things".

Reply:
> What does "fully closed" mean?

As in the connection has been closed, the session is not just `ISHUT`, or `OSHUT`.

> Also below you mention that if the connection has been closed without receiving a CLOSE frame, then this is invoked?!?

Yes, in the event of some connection failure then this will be invoked with some error status code such as 1006.
```
Abnormal Close is a synonym for {@link #NO_CLOSE}, used to indicate a close
condition where no close frame was processed from the remote side.
```

> How can an application distinguish all these cases?

The application can call `session.isOpen()` to know if they can still send messages to the remote endpoint.

I think the main point is that it can attempt to send an of an alternate close status from the one it received. The close does not throw so they don't even really need to check if the session is still open.

Up to you if you want to include any of this info in the documentation, might be too much of implementation details that people don't really need to know.